### ISSUE TITLE
Define how logstash behaves with multiple pipelines feature

### ISSUE THREAD

**Issue by jsvd** [elastic/logstash#6926]
TL;DR Conclusions for 6.x:

1. The multiple pipelines are declared in a YAML file. To make reload semantics easier to understand, this file will be separate from the global `logstash.yml`, called `pipelines.yml`.
2. Starting logstash without arguments now will read pipelines from the yaml file instead of returning error.
3. The three methods of specifying pipelines (pipelines.yml, `-f` and `-e`) will become mutually exclusive between each other.
3.1. This means `-e` will no longer perform concatenation with `-f` (or `pipelines.yml` pipelines).
4. Detecting the absence and autocompletion of input/output sections in the config will be restricted to `-e` (no longer auto completes for `-f`).
5. When reloading is on, `logstash.yml` won't be polled for changes, only `pipelines.yml` will.
6. There will be a tool to generate a `pipelines.yml` for you if you're migrating your `-f` usage from 5.x to 6.x
7. For the Web API, the endpoints that refer to a single pipeline will redirect to `/pipelines/:id` where `:id` will be extracted from `LogStash::SETTINGS.get("pipeline.id")`

--------------------------------------

With the upcoming multiple pipeline targetted at 6.0, it may be interesting to clarify how logstash works when defining and declaring pipelines, either through the cli or configuration files.

Pre MP (multi pipeline), logstash accepts:
  * from the CLI: either -f or -e, or both
  * from the settings.yml: either `config.string` or `path.config` or both

With MP, there is an extra source of configurations: there is a pipelines.yml file where pipelines (their configs and their settings) can be declared. Pipelines in the yaml file also can declare their configurations inline (using `config.string`) or point to files (`path.config`).

Some questions start to pop up:
1. should it be possible to declare pipelines in the pipelines.yml and also use -f and -e too? What about only -e?
2. should running logstash without arguments (`bin/logstash`) now default to using the `pipelines.yml` file? Currently it fails since neither -f nor -e were set.
3. While -e is a very useful tool for experimenting and debugging in Logstash, is it worth creating a line between logstash running in single pipeline mode (`-f`) or with 1 or more pipelines through the `pipelines.yml` file?

Here are my proposals:
1. Eliminate -f so that pipeline descriptions are done through the yaml file
2. Keep -e but make it create an isolated pipeline instead of supporting concatenation with file-based pipelines.
3. Make -e start logstash with just that pipeline, ignoring any pipelines if present in pipelines.yml (with a warning?). This would be kind of a "dev mode" for Logstash.
4. Running logstash without arguments starts the pipelines in pipelines.yml. Make it terminate immediately if pipelines.yml is empty unless `pipeline.reload.automatic` is true.

**Comment by suyograo** [comment]
Some quick thoughts:

Can you define what is in the `pipeline.yml` file? Do we really need it if we can allow users to define pipeline specific configuration in the `logstash.yml` file? My concern is users having to look into too many files for configuration. I'd like to consolidate everything config in one file. So it would be worth it to define what pipeline specific configuration we are allowing. My advice is to make this limited to essential settings for v1. Lesser the better.

To keep things simple, we should remove `-e` altogether. With config reloading, it is now really easy for someone to add `stdin` and `stdout` to debug their config. Also, we should achieve functionality provided by `-e` in a different way using a `simulate` API of sorts. `-e` has been really hard to support with the newer changes in the pipeline. Simulate API would automatically add `stdin` and `stdout` to a specified pipeline and accept events via HTTP.

`-f` is really useful and we should continue to support it since it is the defacto way of starting LS. If `-f` is supplied in the CLI, we should _assume_ a single pipeline mode. We ignore the config in the `logstash.yml`. In other words, `-f` and configuring pipelines in `logstash.yml` is mutually exclusive. `-f` is used for quick way to get started. `logstash.yml` is more persistent, production way.

**Comment by jsvd** [comment]
Thanks for the input, Suyog, my answers:

> Can you define what is in the pipeline.yml file?
Me and PH have already talked about introducing a PipelineSettings class, which is harbors the subset of all settings that are pipeline specific. The pipelines.yml file would then have a list of hashes where the keys are these pipeline settings. Example:

```yaml
- pipeline.id: test
  pipeline.workers: 1
  pipeline.batch.size: 1
  config.string: "input { generator {} } filter { sleep { time => 1 } } output { stdout { codec => plain { format => '1' } } }"
- pipeline.id: nginx
  pipeline.workers: 2
  path.config: "/etc/conf.d/logstash/pipeline_nginx.cfg"
- pipeline.id: businesskpi
  pipeline.batch.size: 10000
  path.config: "/etc/conf.d/logstash/pipeline_kpi.cfg"
```

> To keep things simple, we should remove -e altogether.
I'm a very heavy user of -e so I was a bit surprised by this. We'd be removing the ability of starting pipelines without touching files, which IMO is very useful for quick checking filter/inputs behaviours. As for the simulate API, I believe it is still far off from being released?

> -f is really useful and we should continue to support it since it is the defacto way of starting LS.
For most production scenarios, the start up scripts hide the fact that -f is being used and it is simple to keep the current behaviour by shipping a default `pipelines.yml` with only:

```yaml
- pipeline.id: "main"
  config.path: "/etc/conf.d/logstash/*"
```

**Comment by andrewvc** [comment]
I would also like to keep `-e`. What's the added complexity? Wouldn't it follow the same codepath as `-f` more or less?

**Comment by ph** [comment]
I would like to keep `-e`, and `-f`, what I don't like is the fact that the currently running pipeline is a merged result of `-e` and `-f`+ implicit stdin/stdout added if they are missing.

**Comment by ph** [comment]
I think we should leverage the fact that the yaml files is just a hash representation and since our **pipeline_id** need to be unique and are defined by the user, we should use it as a key. WDYT?

```yaml
pipelines: 
  test:
    pipeline.workers: 1
    pipeline.batch.size: 1
    config.string: "input { generator {} } filter { sleep { time => 1 } } output { stdout { codec => plain { format => '1' } } }"

**Comment by jsvd** [comment]
> I would like to keep -e, and -f, what I don't like is the fact that the currently running pipeline is a merged result of -e and -f+ implicit stdin/stdout added if they are missing.

Removing `-f` would also help break this notion of pipeline concatenation, since logstash would either have -e to run a cli pipeline, or run in "production" with pipelines described in pipelines.yml.

> I think we should leverage the fact that the yaml files is just a hash representation and since our pipeline_id need to be unique and are defined by the user, we should use it as a key. WDYT?

`pipeline.id` is a known setting in logstash, and it becomes confusing to have that pipeline setting but then using a hash in pipelines.yml whose key is this `pipeline.id`. would this mean that `pipeline.id` key would be blacklisted from the hash, i.e. not a "PipelineSetting"?

**Comment by suyograo** [comment]
I take back my comment on removing `-e`. When I wrote that, I was actually thinking of removing the merge behavior that comes with `-e` which is what @ph is referring to. I updated my comments.

We should support both `-e` and `-f` from the CLI, but make it mutually exclusive options. Remove the merge behavior and replace it with the simulate API. 

If a user is in the CLI mode (`-f` or `-e`), we assume single-pipeline. If they want to use multiple pipelines, they would have to specify pipelines in `logstash.yml`. In the CLI mode, we would not look at `logstash.yml` for pipeline definitions. This means LS should be able to start without `-f`.

```
bin/logstash
```

will look at `logstash.yml` for pipeline definitions. If it fails to find any, it should bail

```
bin/logstash -f foo.conf
```

or 

```
bin/logstash -e 'input {stdin {}}'
```
will ignore `logstash.yml`.

**Comment by suyograo** [comment]
Finally, do we need a `pipelines.yml` or can we continue to use `logstash.yml`? Mo files, mo problems.

**Comment by jsvd** [comment]
> Finally, do we need a pipelines.yml or can we continue to use logstash.yml to define pipelines? Mo files, mo problems.

The logic behind a separate file is to make reload semantics simpler: if reload is on, when you change the pipelines.yml file, it will detect changes and reload pipelines. If we only use logstash.yml, we'll only reload if changes happen to the `pipelines` key in this file. All other changes should not trigger a reload, like changes to `http.port` or `config.reload.automatic`.

**Comment by suyograo** [comment]
> The logic behind a separate file is to make reload semantics simpler: if reload is on, when you change the pipelines.yml file, it will detect changes and reload pipelines. 

That's a fair point. `pipelines.yml` makes sense in this case.

@jsvd to throw in more complexity here :), with centralized config, do we ignore `pipelines.yml` completely? Having `pipelines.yml` separated kinda makes that logic easier. For localized config management, use `pipelines.yml`. For centralized config management, use ES as your remote `pipelines.yml`.

**Comment by ph** [comment]
> @jsvd to throw in more complexity here :), with centralized config, do we ignore pipelines.yml completely? Having pipelines.yml separated kinda makes that logic easier. For localized config management, use pipelines.yml. For centralized config management, use ES as your remote pipelines.yml.

In the current implementation, when centralized config is *enabled* all other source loaders are removed and es acts a the source of truth.

**Comment by jsvd** [comment]
closing this as my questions have been answered.
